thoth/common/openshift.py

```diff
@@ -0,0 +1,356 @@
+                "Unable to import OpenShift and Kubernetes packages. Was thoth-common library "
+                "installed with openshift extras?"
+            ) from exc
+
+        # Load in-cluster configuration that is exposed by OpenShift/k8s configuration.
+        config.load_incluster_config()
+
+        self.ocp_client = DynamicClient(client.ApiClient(configuration=client.Configuration()))
+        self.frontend_namespace = frontend_namespace or os.getenv('THOTH_FRONTEND_NAMESPACE')
+        self.middletier_namespace = middletier_namespace or os.getenv('THOTH_MIDDLETIER_NAMESPACE')
+        self.backend_namespace = backend_namespace or os.getenv('THOTH_BACKEND_NAMESPACE')
+        self.infra_namespace = infra_namespace or os.getenv('THOTH_INFRA_NAMESPACE')
+        self.kubernetes_api_url = kubernetes_api_url or \
+                                  os.getenv('KUBERNETES_API_URL', 'https://kubernetes.default.svc.cluster.local')
+        self.kubernetes_verify_tls = bool(kubernetes_verify_tls or os.getenv('KUBERNETES_VERIFY_TLS', True))
+        self.openshift_api_url = openshift_api_url or \
+                                 os.getenv('OPENSHIFT_API_URL', 'https://openshift.default.svc.cluster.local')
+        self._token = token
+
+    @property
+    def token(self):
+        """Access service account token mounted to the pod."""
+        if self._token is None:
+            self._token = get_service_account_token()
+
+        return self._token
+
+    @staticmethod
+    def _set_env_var(template: dict, **env_var):
+        """Set environment in the given template."""
+        for env_var_name, env_var_value in env_var.items():
+            for entry in template['spec']['containers'][0]['env']:
+                if entry['name'] == env_var_name:
+                    entry['value'] = env_var_value
+                    break
+            else:
+                template['spec']['containers'][0]['env'].append(
+                    {'name': env_var_name, 'value': str(env_var_value)}
+                )
+
+    @staticmethod
+    def _set_template_parameters(template: dict, **parameters: object) -> None:
+        """Set parameters in the template - replace existing ones or append to parameter list if not exist.
+
+        >>> _set_template_parameters(template, THOTH_LOG_ADVISER='DEBUG')
+        """
+        if 'parameters' not in template:
+            template['parameters'] = []
+
+        for parameter_name, parameter_value in parameters.items():
+            for entry in template['parameters']:
+                if entry['name'] == parameter_name:
+                    entry['value'] = str(parameter_value)
+                    break
+            else:
+                template['parameters'].append({
+                    'name': parameter_name,
+                    'value': str(parameter_value)
+                })
+
+    def run_sync(self, force_analysis_results_sync: bool = False, force_solver_results_sync: bool = False) -> str:
```

Comment:
this whole section of preparing the pod/job template looks complicated, why is that?

Reply:
It runs job and reuses all the configuration of cronjob.